HelpSource/Classes/ArrayedCollection.schelp

```diff
@@ -420,7 +420,7 @@ code::
 ::
 
 method::perfectShuffle
-Returns a copy of the receiver with its items split into two equal halves, then reconstructed by interleaving the two halves. note::use an even number of item pairs in order to not loose any items in the shuffle.::
+Returns a copy of the receiver with its items split into two equal halves, then reconstructed by interleaving the two halves. note::the size of the collection should be even, otherwise an item will be lost in the shuffle.::
```

Comment:
Is it worth specifying which item gets lost? From a quick test it looks like the first item in the "second set" of items after the split.

Reply:
Sure, that sounds like a good idea to me. I might suggest the following:

```
note::the size of the collection should be even, otherwise the item directly in the middle of the collection will be lost in the shuffle.::
```

Interestingly, though, when the collection has size 1, perfectShuffle returns the receiver and nothing is lost. But this is probably not worth mentioning in the note.

Reply:
That sounds good to me!

Reply:
Great! Should I make another commit on the same branch and do another PR? I'm kind of a GitHub novice and am not totally clear on the correct workflow.

Reply:
You can just make another commit and either squash it with the previous and force push, or just make a second commit and we'll squash them when it's merged.

Reply:
> Great! Should I make another commit on the same branch and do another PR? I'm kind of a GitHub novice and am not totally clear on the correct workflow.

BTW I recently made a [tutorial](https://github.com/supercollider/supercollider/wiki/Contributing-to-SC-using-a-GUI) on how to contribute to SC using GitHub Desktop, which makes updating, reordering, and squashing commits very easy.

Reply:
Oh, nice! I will check that out. I've been using the command line (which I actually do like), so I might stick with that for this PR since it's such a minor change. I'm planning to do the following (amending the previous commit seems appropriate here), but I have a history of messing stuff up. Can you confirm the following would be appropriate?

```
$ git branch topic/ArrayedCollection-perfectShuffle-note
# (make and save discussed changes)
$ git add .
$ git commit --amend
$ git push --set-upstream origin topic/ArrayedCollection-perfectShuffle-note
# (do another PR on GitHub)
```

Reply:
Sure! 

Disclaimer: I'm not all that deep into git workflow, so these are best guesses :) 

I don't know offhand the specific workflow for `--amend`, but based on a quick look at [this tutorial](https://www.atlassian.com/git/tutorials/rewriting-history), I think you may want to do something like:
```shell
# switch to your feature branch (you may already be on it)
$ git checkout topic/ArrayedCollection-perfectShuffle-note
# (make and save discussed changes)
$ git add .
# (amend the previous commit without changing it's commit message)
$ git commit --amend --no-edit
$ git push
```
The differences are that it uses `checkout` to switch to the feature branch you're making changes on.
Then the `--no-edit` flag indicates to use the same commit message as the commit being amended. (Alternatively `git commit --amend -m "an updated commit message"`)
Then I think you wouldn't need to `--set-upstream` when pushing, because you've already opened a PR on your feature branch that lives on your remote origin repo (in your GitHub account), so it's "linked" to the PR and anything you push to that remote branch will automatically update in the PR.

In any case, I _think_ amending a commit with actual changes to files (as opposed to amending a commit _message_ only) will create a new hash for that entry in the timeline, which will overwrite your previous commit. This is the same result as if you just made a second commit then squashed the two into one (creating a new hash) then force-push it.

Don't worry about messing anything up, no one is working on this branch with you so you can make all the mess you want, and multiple commits can be squashed into one by the merging reviewer. So experiment away!

Reply:
Ok, thanks for the help. Unfortunately, I am having some trouble. When I try to push, I get this:

```
 ! [rejected]            topic/ArrayedCollection-perfectShuffle-note -> topic/ArrayedCollection-perfectShuffle-note (non-fast-forward)
error: failed to push some refs to 'https://github.com/elifieldsteel/supercollider'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

So this seems to suggest I need to `$ git pull` first. When I do, I get:

```
fatal: Not possible to fast-forward, aborting.
```

I've tried all three git config pull options (merge, rebase, fast-forward only), but I get the same message. I am pretty clueless here. Any idea how to get myself out of this mess?

Reply:
Also, when I `$ git status`, I get:

```
On branch topic/ArrayedCollection-perfectShuffle-note
Your branch and 'origin/topic/ArrayedCollection-perfectShuffle-note' have diverged,
and have 1 and 1 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)
```

Which again suggests I need to pull to bring things up-to-date. But, as mentioned, a pull attempt gives a 'fatal' message.

Reply:
So this is related to the aspect that I mentioned—that the hashes (IDs) of the remote and corresponding (amended) local commit are now different.

Because the _content_ of the amended commit has changed it is was assigned a new ID, essentially making it new commit altogether. Locally, it has _replaced_ its amended counterpart—so your local git history shows just one (updated) commit. When you try to push that amended commit to your remote repo, your intention is to replace the original commit you pushed previously. Git won't let you do this implicitly (that's a good thing). 

To step back a moment—conceptually, your remote origin repo (on GitHub) must be _identical_ to the one you have locally, they should be mirrors of one another. When first cloned your repo, or when you pull from the remote to your local repo, they are identical: your local copy is brought "in sync" with the remote. When you then make a change locally (and commit it), you're "ahead" of the remote by one commit. To bring your remote in sync with your local repo, you `git push` your local changes/commits to the remote; your remote's history is brought "in sync" with your local repo. You can continue working this way as long as all of your changes happen locally, and they're pushed to the remote ("one way" so to speak).

By using `--amend`, you've changed the content of the most recent local commit. In doing so, git assigns it a new ID. So even though it is _conceptually_ one commit (it was just "updated"), in practice, you've made a _new_ commit (with a new ID) that replaces the former one. 

However, your remote still has the "old" commit (with the old ID). So with the error it's saying "hey, before you can push local changes to the remote, both repos need to have an identical history prior to your new changes". But you're intention is to say "actually, I know they're out of sync, but I want to overwrite the remote content with what I have locally anyway". To do that, you need to use 
```shell
git push -f
```
i.e. _force_ push.

You may be asked for a password or given some cautionary prompt, but go for it.

Reply:
Thanks so much for having the patience to help me through this issue. It looks like the force push worked, and I can see the updated changes here in the open PR. So I think there's nothing left for me to do, right?

Reply:
Yep, you got it! Thanks for following through :)